src/mongo/MongoShell.ts

```diff
@@ -0,0 +1,179 @@
+		script = convertToSingleLine(script);
+
+		let stdOut = "";
+		const sentinel = createSentinel();
+
+		let disposables: vscode.Disposable[] = [];
+		try {
+			let result = await new Promise<string>(async (resolve, reject) => {
+				try {
+					startScriptTimeout(this._timeoutSeconds, reject);
+
+					// Hook up events
+					disposables.push(
+						this._process.onStdOut(text => {
+							stdOut += text;
+							let { text: stdOutNoSentinel, removed } = removeSentinel(stdOut, sentinel);
+							if (removed) {
+								// The sentinel was found, which means we are done.
+
+								// Change the "type 'it' for more" message to one that doesn't ask users to type anything,
+								//   since we're not currently interactive like that.
+								// CONSIDER: Ideally we would allow users to click a button to iterate through more data,
+								//   or even just do it for them
+								stdOutNoSentinel = stdOutNoSentinel.replace(mongoShellMoreMessage, extensionMoreMessage);
+
+								resolve(stdOutNoSentinel);
+							}
+						}));
+					disposables.push(
+						this._process.onStdErr(text => {
+							// Mongo shell only writes to STDERR for errors relating to starting up. Script errors go to STDOUT.
+							//   So consider this an error.
+							// (It's okay if we fire this multiple times, the first one wins.)
+							reject(wrapCheckOutputWindow(text.trim()));
+						}));
+					disposables.push(
+						this._process.onError(error => {
+							reject(error);
+						}));
+
+					// Write the script to STDIN
+					if (script) {
+						this._process.writeLine(script);
+					}
+
+					// Mark end of result by sending the sentinel wrapped in quotes so the console will spit
+					// it back out as a string value after it's done processing the script
+					let quotedSentinel = `"${sentinel}"`;
+					this._process.writeLine(quotedSentinel); // (Don't display the sentinel)
+
+				} catch (error) {
+					// new Promise() doesn't seem to catch exceptions in an async function, we need to explicitly reject it
+
+					if ((<{ code?: string }>error).code === 'EPIPE') {
+						// Give a chance for start-up errors to show up before rejecting with this more general error message
+						await delay(500);
+						error = new Error("The process exited prematurely.");
+					}
+
+					reject(wrapCheckOutputWindow(error));
+				}
+			});
+
+			return result.trim();
+		}
+		finally {
+			// Dispose event handlers
+			for (let d of disposables) {
+				d.dispose();
+			}
+		}
+	}
+}
+
+function startScriptTimeout(timeoutSeconds: number | 0, reject: (unknown) => void): void {
+	if (timeoutSeconds > 0) {
+		setTimeout(
+			() => {
+				reject(timeoutMessage);
+			},
+			timeoutSeconds * 1000);
+	}
+}
+
+function convertToSingleLine(script: string): string {
+	return script.split(os.EOL)
+		.map(line => line.trim())
+		.join('')
+		.trim();
+
+}
+
+function removeSentinel(text: string, sentinel: string): { text: string; removed: boolean } {
```

Comment:
Can there be more than one sentinel in the text? This removes just the first occurrence of sentinel.